File: app/src/main/java/com/example/notificationapp/view/activities/SignUpActivity.java
package com.example.notificationapp.view.activities;

import static android.widget.Toast.makeText;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.example.notificationapp.R;
import com.example.notificationapp.data.network.api.RetrofitAccessObject;
import com.example.notificationapp.data.network.model.UserModel;
import com.example.notificationapp.data.network.model.UserResponse;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SignUpActivity extends AppCompatActivity {

    private TextInputEditText mInputEmail, mInputPassword, mInputName, mInputMobile, mInputRegNo;
    private TextView mLoginTV;
    private AutoCompleteTextView mATVYear, mATVCourse;
    private String mCourse, mYear;
    private Button mSignUpBtn;
    private ConstraintLayout mParent;
    private final List<String> itemsCourse = Arrays.asList("B.Tech", "M.Tech", "MBA", "MCA", "PhD");
    private final List<Integer> itemsYear = Arrays.asList(2023, 2024, 2025, 2026);

    private FirebaseAuth mAuth;
    private static final String TAG = "SignUpActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_signup);
        Objects.requireNonNull(getSupportActionBar()).hide();
        mAuth = FirebaseAuth.getInstance();
        setReferences();
        setListeners();


        ArrayAdapter<String> adapterCourse = new ArrayAdapter<>(this, R.layout.list_item, itemsCourse);
        mATVCourse.setAdapter(adapterCourse);

        mATVCourse.setOnItemClickListener((adapterView, view, position, id) -> {
            mCourse = adapterView.getItemAtPosition(position).toString();
        });

        ArrayAdapter<Integer> adapterYear = new ArrayAdapter<>(this, R.layout.list_item, itemsYear);
        mATVYear.setAdapter(adapterYear);

        mATVCourse.setOnItemClickListener((adapterView, view, position, id) -> {
            mYear = adapterView.getItemAtPosition(position).toString();
        });

        mLoginTV.setOnClickListener(view -> {
            startActivity(new Intent(getApplicationContext(), LoginActivity.class));
            finish();
        });
    }

    private void setListeners() {
        mSignUpBtn.setOnClickListener(view -> signUpUser());
    }

    private void signUpUser() {
        String emailText = mInputEmail.getText().toString();
        String passwordText = mInputPassword.getText().toString();
        String mobileText = mInputMobile.getText().toString();
        String nameText = mInputName.getText().toString();
        String regNoText = mInputRegNo.getText().toString();
        String courseText = mATVCourse.getText().toString();
        String yearText = mATVYear.getText().toString();

        if (!validate()) return;

        mAuth.createUserWithEmailAndPassword(emailText, passwordText)
                .addOnCompleteListener(this, task -> {
                    if (task.isSuccessful()) {
                        FirebaseUser user = mAuth.getCurrentUser();

                        user.sendEmailVerification().addOnCompleteListener(new OnCompleteListener<Void>() {
                            @Override
                            public void onComplete(@NonNull Task<Void> task) {
                                if (task.isSuccessful()) {
                                    Log.d(TAG, "Email sent.");
                                    user.getIdToken(true).addOnSuccessListener(result -> {
                                        String idToken = result.getToken();
                                        UserModel userModel = new UserModel();
                                        userModel.setEmail(emailText);
                                        userModel.setName(nameText);
                                        userModel.setCourse(courseText);
                                        userModel.setGraduationYear(yearText);
                                        userModel.setPersonalEmail(emailText);
                                        userModel.setPhoneNumber(mobileText);
                                        userModel.setRegistrationNumber(regNoText);

                                        RetrofitAccessObject.getRetrofitAccessObject().saveUser(idToken, userModel).enqueue(new Callback<UserResponse>() {
                                            @Override
                                            public void onResponse(Call<UserResponse> call, Response<UserResponse> response) {
                                                if (response.body() != null) {
                                                    makeText(SignUpActivity.this,
                                                            "Registered Successfully, Please Verify Your Account and login!",
                                                            Toast.LENGTH_SHORT).show();
                                                    goToLogin();
                                                }
                                            }

                                            @Override
                                            public void onFailure(Call<UserResponse> call, Throwable t) {
                                                retry(call.toString());
                                                mSignUpBtn.setEnabled(true);
                                            }
                                        });
                                    });
                                }
                            }
                        });

                    } else {
                        makeText(SignUpActivity.this, Objects.requireNonNull(task.getException()).getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void setReferences() {
        mInputEmail = findViewById(R.id.et_email);
        mInputPassword = findViewById(R.id.et_password);
        mATVYear = findViewById(R.id.et_grad_year);
        mATVCourse = findViewById(R.id.et_course);
        mInputMobile = findViewById(R.id.et_mobile);
        mInputName = findViewById(R.id.et_username);
        mInputRegNo = findViewById(R.id.et_reg_no);
        mSignUpBtn = findViewById(R.id.signup_btn);
        mParent = findViewById(R.id.parent);
        mLoginTV = findViewById(R.id.login);
    }

    private boolean validate() {
        String emailText = mInputEmail.getText().toString();
        String passwordText = mInputPassword.getText().toString();
        String mobileText = mInputMobile.getText().toString();
        String nameText = mInputName.getText().toString();
        String regNoText = mInputRegNo.getText().toString();

        if (emailText.equals("")) {
            mInputEmail.requestFocus();
            return false;
        }
        if (passwordText.equals("")) {
            mInputPassword.requestFocus();
            return false;
        }
        if (nameText.equals("")) {
            mInputName.requestFocus();
            return false;
        }
        if (mobileText.equals("")) {
            mInputMobile.requestFocus();
            return false;
        }
        if (regNoText.equals("")) {
            mInputRegNo.requestFocus();
            return false;
        }
        String s = emailText;
        String compare = "@mnnit.ac.in";
        int j = compare.length() - 1;
        boolean check = false;
        int n = emailText.length() - 1;

        check = false;

        //helperTextForEmail.text = "Please enter the valid e-mail address"

        while (n >= 0 && j >= 0) {

            if (s.charAt(n) != compare.charAt(j)) {
                break;
            }

            n--;
            j--;
        }
        if (j == -1 && compare.length() < s.length()) {
            check = true;
            //helperTextForEmail.text = "Valid Email entered!"
        } else {
            Log.d(TAG, "Invalid");
            makeText(getApplicationContext(), "Please Enter G-Suite ID.", Toast.LENGTH_LONG).show();
        }
        return check;
    }

    private void goToLogin() {
        Intent intent = new Intent(this, LoginActivity.class);
        startActivity(intent);
        finish();
    }

    private void retry(String message) {
        Snackbar.make(mParent, message, Snackbar.LENGTH_LONG).show();
        mSignUpBtn.setEnabled(true);
    }
}

File: app/src/main/java/com/example/notificationapp/view/activities/SignUpActivity.kt
package com.example.notificationapp.view.activities

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.notificationapp.R
import com.example.notificationapp.data.network.API
import com.example.notificationapp.data.network.UserModel
import com.example.notificationapp.databinding.ActivitySignupBinding
import com.example.notificationapp.isNotValidDomain
import com.google.android.gms.tasks.Task
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GetTokenResult

class SignUpActivity : AppCompatActivity() {

    companion object {
        private const val TAG = "SignUpActivity"
    }

    private lateinit var binding: ActivitySignupBinding

    private lateinit var mAuth: FirebaseAuth

    private var mCourse = ""
    private var mYear = ""
    private val itemsCourse = listOf("B.Tech", "M.Tech", "MBA", "MCA", "PhD")
    private val itemsYear = listOf(2023, 2024, 2025, 2026)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySignupBinding.inflate(layoutInflater)
        setContentView(binding.root)
        supportActionBar?.hide()

        mAuth = FirebaseAuth.getInstance()
        setListeners()

        val adapterCourse = ArrayAdapter(this, R.layout.list_item, itemsCourse)
        binding.etCourse.setAdapter(adapterCourse)
        binding.etCourse.onItemClickListener = OnItemClickListener { adapterView: AdapterView<*>, _: View?, position: Int, _: Long ->
            mCourse = adapterView.getItemAtPosition(position).toString()
        }
        val adapterYear = ArrayAdapter(this, R.layout.list_item, itemsYear)
        binding.etGradYear.setAdapter(adapterYear)
        binding.etGradYear.onItemClickListener = OnItemClickListener { adapterView: AdapterView<*>, _: View?, position: Int, _: Long ->
            mYear = adapterView.getItemAtPosition(position).toString()
        }
        binding.login.setOnClickListener {
            startActivity(Intent(applicationContext, LoginActivity::class.java))
            finish()
        }
    }

    private fun setListeners() {
        binding.signupBtn.setOnClickListener { signUpUser() }
    }

    private fun signUpUser() {
        val emailText = binding.etEmail.text?.toString() ?: ""
        val passwordText = binding.etPassword.text?.toString() ?: ""
        val regNoText = binding.etRegNo.text?.toString() ?: ""
        val mobileText = binding.etMobile.text?.toString() ?: ""
        val nameText = binding.etUsername.text?.toString() ?: ""
        val courseText = binding.etCourse.text?.toString() ?: ""
        val yearText = binding.etGradYear.text?.toString() ?: ""
        if (!validate(emailText, passwordText, regNoText, mobileText, nameText, courseText, yearText)) return

        binding.signupBtn.isEnabled = false
        mAuth.createUserWithEmailAndPassword(emailText, passwordText)
            .addOnCompleteListener(this) { createUserTask: Task<AuthResult?> ->
                if (!createUserTask.isSuccessful) {
                    Toast.makeText(this@SignUpActivity, createUserTask.exception?.message ?: "Null", Toast.LENGTH_SHORT).show()
                    return@addOnCompleteListener
                }
                val user = mAuth.currentUser
                if (user == null) {
                    Toast.makeText(this@SignUpActivity, "Error: User null despite sign up", Toast.LENGTH_SHORT).show()
                    return@addOnCompleteListener
                }
                user.sendEmailVerification().addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        Log.d(TAG, "Email sent.")
                        user.getIdToken(true).addOnSuccessListener { result: GetTokenResult ->
                            val idToken = result.token
                            val userModel = UserModel(nameText, regNoText, yearText, courseText, emailText, emailText, mobileText)
                            API.saveUser(idToken, userModel, {
                                Toast.makeText(
                                    this@SignUpActivity,
                                    "Registered Successfully, Please Verify Your Account and login!",
                                    Toast.LENGTH_SHORT
                                ).show()
                                goToLogin()
                            }) { retry("$it: Error Signing Up") }
                        }
                    }
                }
            }
    }

    private fun validate(
        emailText: String,
        passwordText: String,
        regNoText: String,
        mobileText: String,
        nameText: String,
        courseText: String,
        yearText: String
    ): Boolean {
        if (emailText.isEmpty()) {
            binding.etEmail.requestFocus()
            return false
        }
        if (emailText.isNotValidDomain()) {
            Toast.makeText(this, "Please use G-Suite ID", Toast.LENGTH_SHORT).show()
            binding.etEmail.requestFocus()
            return false
        }

        if (regNoText.isEmpty()) {
            binding.etRegNo.requestFocus()
            return false
        }
        if (!"^\\d{8}$".toRegex().containsMatchIn(regNoText)) {
            Toast.makeText(this, "Invalid registration number", Toast.LENGTH_SHORT).show()
            binding.etRegNo.requestFocus()
            return false
        }

        if (passwordText.isEmpty()) {
            binding.etPassword.requestFocus()
            return false
        }
        if (courseText.isEmpty()) {
            binding.etCourse.requestFocus()
            return false
        }
        if (yearText.isEmpty()) {
            binding.etGradYear.requestFocus()
            return false
        }
        if (nameText.isEmpty()) {
            binding.etUsername.requestFocus()
            return false
        }
        if (mobileText.isEmpty()) {
            binding.etMobile.requestFocus()
            return false
        }
        return true
    }

    private fun goToLogin() {
        val intent = Intent(this, LoginActivity::class.java)
        startActivity(intent)
        finish()
    }

    private fun retry(message: String) {
        Snackbar.make(binding.parent, message, Snackbar.LENGTH_LONG).show()
        binding.signupBtn.isEnabled = true
    }
}